server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -315,25 +316,41 @@ public boolean offer(E e) {
      * A handler for rejected tasks that adds the specified element to this queue,
      * waiting if necessary for space to become available.
      */
-    static class ForceQueuePolicy implements XRejectedExecutionHandler {
+    static class ForceQueuePolicy extends EsRejectedExecutionHandler {
+
+        /**
+         * This flag is used to indicate if {@link Runnable} should be rejected once the thread pool is shutting down, ie once
+         * {@link ThreadPoolExecutor#shutdown()} has been called. Scaling thread pools are expected to always handle tasks rejections, even
+         * after shutdown or termination, but it's not the case of all existing thread pools so this flag allows to keep the previous
+         * behavior.
+         */
+        private final boolean rejectAfterShutdown;
+
+        /**
+         * @param rejectAfterShutdown indicates if {@link Runnable} should be rejected once the thread pool is shutting down
+         */
+        ForceQueuePolicy(boolean rejectAfterShutdown) {
+            this.rejectAfterShutdown = rejectAfterShutdown;
+        }
 
         @Override
-        public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+        public void rejectedExecution(Runnable task, ThreadPoolExecutor executor) {
             try {
+                final BlockingQueue<Runnable> queue = executor.getQueue();
                 // force queue policy should only be used with a scaling queue
-                assert executor.getQueue() instanceof ExecutorScalingQueue;
-                executor.getQueue().put(r);
+                assert queue instanceof ExecutorScalingQueue;
+                queue.put(task);
+                if (rejectAfterShutdown) {
+                    if (executor.isShutdown() && executor.remove(task)) {
```

Comment:
nit: I would find this slightly more logical as:

```suggestion
                if (rejectAfterShutdown && executor.isShutdown()) {
                    if (executor.remove(task)) {
```

since that seems to be the special case. With the remove being mutating I like that in its own condition to not think about and/or and order of evaluation, but could also collapse into one if statement.

Reply:
This is more logical I agree. (I collapsed the statements)